pkg/preparer/status.go

```diff
@@ -0,0 +1,91 @@
+package preparer
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+
+	"github.com/square/p2/pkg/logging"
+)
+
+type StatusServer struct {
+	listener net.Listener
+	server   *http.Server
+	logger   *logging.Logger
+	Exit     chan error
+}
+
+func (s *StatusServer) Close() error {
+	return s.listener.Close()
+}
+
+var NoServerConfigured = fmt.Errorf("No status server was configured")
+
+func NewStatusServer(statusPort int, statusSocket string, logger *logging.Logger) (*StatusServer, error) {
+	server := http.Server{}
+	statusServer := &StatusServer{
+		server: &server,
+		logger: logger,
+		Exit:   make(chan error),
+	}
+	var listener net.Listener
+	var err error
+
+	if statusPort != 0 {
+		listener, err = statusServer.listenOnPort(statusPort)
+		if err != nil {
+			return nil, err
+		}
+	} else if statusSocket != "" {
+		listener, err = statusServer.listenOnSocket(statusSocket)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		return nil, NoServerConfigured
+	}
+
+	statusServer.listener = listener
+
+	return statusServer, nil
+}
+
+func (s *StatusServer) Serve() {
+	defer s.Close()
+	mux := http.NewServeMux()
+	mux.HandleFunc("/_status", func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintf(w, "p2-preparer OK")
+	})
+
+	s.server.Handler = mux
+	err := s.server.Serve(s.listener)
+	s.logger.WithError(err).Warnln("Status server exited!")
+	s.Exit <- err
+	close(s.Exit)
```

Comment:
Currently unused but could be retrieved and used to kill the preparer on failure.